### Semantic Kernel (SK) + GPT‑5 — BFSI


This notebook introduces Semantic Kernel (SK) — a model‑agnostic SDK for building, orchestrating, and deploying AI agents and multi‑agent systems.

**Default provider:** OpenAI GPT‑5 (set via `OPENAI_API_KEY`)

### 1️⃣ What is Semantic Kernel?
**Semantic Kernel (SK)** is a **model‑agnostic SDK** that enables developers to build, orchestrate, and deploy AI agents and multi‑agent systems.

**Key Features**:
- 🔄 Model Flexibility (OpenAI, Azure, Hugging Face, NVIDIA, local LLMs)
- 🧠 Agent Framework for modular AI assistants
- 🤝 Multi‑Agent Orchestration (specialist collaboration)
- 🧩 Plugin Ecosystem (native functions, prompts, OpenAPI, MCP)
- 🗂️ Vector DB support (Azure AI Search, Elasticsearch, Chroma, etc.)
- 🪄 Multimodal (text, vision, audio)
- 🏢 Enterprise‑ready: Observability, Security, Stable APIs

## 2️⃣ Why Semantic Kernel?
- **Unify** access to multiple LLM providers
- **Compose** prompts, functions, and APIs as tools
- **Govern** via filters and telemetry for compliance
- **Scale** from simple chat to enterprise multi‑agent workflows
- **Ship** faster with modular, reusable agents

## 3️⃣ Benefits for BFSI
- 🧾 Compliance‑aware prompts and guardrails
- 💡 Explainable outputs and traceability
- 🔒 Secure for sensitive data (on‑prem/local options)
- ⚙️ Fast development of KYC, Risk, and Compliance copilots

## 4️⃣ Environment Setup
Install the SDK and connectors.

In [ ]:
!pip install -U semantic-kernel langchain-openai pydantic

In [ ]:
import os
os.environ["OPENAI_API_KEY"] = "sk-..."  # Replace with your key
print("✅ Environment configured for OpenAI GPT-5")

## 5️⃣ Quickstart: Hello SK Agent

In [55]:
%%writefile semantic_kernel_gpt5_bfsi.py

import os
import asyncio
from semantic_kernel.agents import ChatCompletionAgent
from semantic_kernel.connectors.ai.open_ai import OpenAIChatCompletion

async def main():
    agent = ChatCompletionAgent(
        service=OpenAIChatCompletion(ai_model_id="gpt-5", api_key=os.getenv("OPENAI_API_KEY")),
        name="SK-Assistant",
        instructions="You are a helpful BFSI assistant. Keep responses concise and compliant."
    )
    resp = await agent.get_response(messages="How Semantic Kernel can be used in BFSI domain.")
    print(resp.content)

asyncio.run(main())

Overwriting semantic_kernel_gpt5_bfsi.py


In [56]:
!python semantic_kernel_gpt5_bfsi.py

Semantic Kernel (SK) is an orchestration SDK for building AI “copilots” that can reason over your data and safely call enterprise systems. In BFSI, it’s useful wherever you need grounded, auditable AI that composes LLMs with deterministic services and data.

Where it fits in BFSI
- Retail/consumer banking
  - Conversational banking assistant: balances, transactions, card controls, dispute intake, spending insights via SK functions that wrap core banking APIs.
  - Branch/contact-center agent assist: real-time call summarization, next-best-action, knowledge lookup across products, fees, procedures with RAG.
- Commercial/wholesale banking
  - Credit memo drafting and covenant monitoring: RAG over financials and policies, tool calls to spreading engines and risk calculators, audit trail of sources.
  - Payment investigations: ingest SWIFT/ISO 20022 messages, graph lookups, compile investigation notes, generate client communications.
- Wealth/asset management
  - Portfolio Q&A: RAG over res

Semantic Kernel (SK) is an orchestration SDK for building AI “copilots” that can reason over your data and safely call enterprise systems. In BFSI, it’s useful wherever you need grounded, auditable AI that composes LLMs with deterministic services and data.

Where it fits in BFSI
- Retail/consumer banking
  - Conversational banking assistant: balances, transactions, card controls, dispute intake, spending insights via SK functions that wrap core banking APIs.
  - Branch/contact-center agent assist: real-time call summarization, next-best-action, knowledge lookup across products, fees, procedures with RAG.
- Commercial/wholesale banking
  - Credit memo drafting and covenant monitoring: RAG over financials and policies, tool calls to spreading engines and risk calculators, audit trail of sources.
  - Payment investigations: ingest SWIFT/ISO 20022 messages, graph lookups, compile investigation notes, generate client communications.
- Wealth/asset management
  - Portfolio Q&A: RAG over research, statements, and house views; call risk/valuation engines for scenarios; produce compliant client-ready summaries.
  - Advisor copilot: suitability checks, product comparisons, meeting notes, follow-ups; entitlements-aware content retrieval.
- Insurance
  - FNOL intake and claims triage: entity extraction, policy lookup, coverage checks, draft correspondence; route high-risk cases to human.
  - Underwriting assistant: summarize broker submissions, classify risk, check guidelines, request missing data; integrate with pricing models.
- Risk, compliance, and operations
  - KYC/AML alert triage: summarize alerts, fetch KYC docs, screen hits, produce analyst-ready narratives and SAR drafts with citations.
  - Policy and regulatory Q&A: RAG over internal policies, Basel/IFRS guidance; generate audit-ready rationales with provenance.
  - Model documentation and report drafting: assemble regulatory reports and controls evidence from source systems.

How SK helps (capabilities mapped to BFSI needs)
- Tool/Function orchestration: Wrap core systems (core banking, LOS, claims, pricing, screening, payment rails) as SK functions so the LLM can call them with strict allow-lists and parameter validation.
- Grounded answers with Memory/RAG: Use SK “memory” over vector stores (e.g., enterprise search) to cite policies, procedures, research, product docs; reduce hallucination with retrieval and source citations.
- Planning and multi-step workflows: Use SK planners or explicit chains to break a user goal into safe steps (e.g., “analyze statement → check limits → draft message”), mixing LLM and non-LLM steps.
- Multi-agent patterns: Separate agents for customer service, risk, or compliance that collaborate via SK while enforcing segregation of duties and tool access.
- Guardrails and governance: Content filters, prompt/response redaction for PII, tool call allow-lists, human-in-the-loop checkpoints, and comprehensive logging/traceability.
- Entitlements-aware retrieval: Inject user identity/roles to filter retrieved documents and data, honoring least-privilege and data residency.
- Observability and evaluation: Capture prompts, tool calls, and sources for audit; run offline evals on accuracy, safety, and bias aligned to model risk management.

Typical architecture pattern
- Client (app, IVR, chat) → SK host service (per-use-case kernel).
- Models via approved providers (e.g., enterprise LLM endpoint) with content safety filters.
- SK functions (plugins) for:
  - Data: enterprise search/vector DB; data lakes; policy repositories.
  - Systems: core banking/claims/LOS/CRM/KYC, pricing/risk engines, payments, case mgmt.
  - Utilities: PII redaction, OCR, translation, compliance checkers.
- Planner/Orchestrator: chooses functions, manages multi-step flows, enforces guardrails.
- Security: enterprise authZ (RBAC/ABAC), secrets vault, network isolation, encryption.
- Telemetry: logging, tracing, evaluations; storage for citations and decisions.

Example blueprints
- AML alert triage
  1) Retrieve alert details and related transactions (function).
  2) RAG over KYC files, prior cases, typologies (memory).
  3) Summarize and assess risk indicators; propose next actions (LLM).
  4) Draft SAR narrative with citations; route for human approval (function + checkpoint).
- Claims intake (insurance)
  1) Capture FNOL via chat/voice; extract entities (LLM).
  2) Validate policy and coverages (function).
  3) Detect potential fraud signals via rules/ML service (function).
  4) Generate claimant comms and internal tasks; attach evidence and sources (LLM + functions).
- Credit memo drafting
  1) Pull financials and spreads (function).
  2) RAG over sector outlook and internal policy (memory).
  3) Generate draft memo with covenant checks and cited sources (LLM).
  4) Gate for analyst review and sign-off (checkpoint).

Implementation steps
- Pick a narrow, high-value use case with clear human-in-the-loop.
- Inventory systems to expose as SK functions and define strict schemas/allow-lists.
- Build a curated knowledge index; chunking, metadata, and entitlements filtering.
- Design prompts with tool-use instructions and injection defenses; add content/PII filters.
- Define checkpoints for actions that affect money, customers, or compliance.
- Evaluate offline (accuracy, groundedness, safety) and online (quality, handle time); document per model risk policy (e.g., SR 11-7).
- Pilot with audit logging and monitoring; iterate, then scale.

Risk and compliance best practices
- Ground every material claim with retrieved sources; show citations.
- Never let the model execute monetary or regulatory actions directly; require deterministic services and human approval.
- Minimize and mask PII; use enterprise LLM endpoints with no data retention; enforce data residency.
- Robust prompt-injection defenses; validate/normalize all tool inputs.
- Full audit trail of prompts, tool calls, data sources, and human approvals; retention per policy.
- Regular evaluations for drift, bias, and safety; change control for prompts and tools.

Outcome
Using SK, BFSI firms can deliver compliant, auditable AI copilots that reduce handle time, improve quality and consistency, and safely integrate LLM reasoning with existing systems and controls.

### Example: KYC Summarizer with PII Redaction

In [57]:
%%writefile semantic_kernel_ex2.py

import os
import re, asyncio
from semantic_kernel.agents import ChatCompletionAgent
from semantic_kernel.connectors.ai.open_ai import OpenAIChatCompletion

def redact_pii(text):
    text = re.sub(r'[\w\.-]+@[\w\.-]+', '[REDACTED_EMAIL]', text)
    text = re.sub(r'\b(\+?\d[\d\-\s]{7,})\b', '[REDACTED_PHONE]', text)
    return text

async def kyc_summary(raw_text):
    clean = redact_pii(raw_text)
    agent = ChatCompletionAgent(
        service=OpenAIChatCompletion("gpt-5",api_key=os.getenv("OPENAI_API_KEY")),
        name="KYC-Agent",
        instructions="Summarize KYC rationale for banking in 2 lines. Mask all PII."
    )
    resp = await agent.get_response(messages=clean)
    print(resp.content)

asyncio.run(kyc_summary("Ajay, email ajay@gktcs.com, phone +91-98765-43210 asks why KYC is required."))

Overwriting semantic_kernel_ex2.py


In [19]:
!python semantic_kernel_ex2.py

KYC verifies customer identities to meet legal/regulatory requirements and prevent money laundering, fraud, and terrorist financing. 
It protects banks and customers by reducing financial-crime risk and ensuring safer, transparent, and traceable transactions.


### Example 2: EMI Calculator Tool (Kernel Plugin)

In [82]:
%%writefile semantic_kernel_emi.py

import asyncio
import os
from typing import Annotated
from semantic_kernel import Kernel
from semantic_kernel.connectors.ai.open_ai import OpenAIChatCompletion
from semantic_kernel.functions.kernel_function_decorator import kernel_function


@kernel_function(
    name="calculate_emi",
    description="Compute EMI (Equated Monthly Instalment) for a loan."
)
def calculate_emi(
    principal: Annotated[float, "Loan principal in INR"],
    annual_rate: Annotated[float, "Annual interest rate in %"],
    tenure_years: Annotated[float, "Loan tenure in years"]
) -> float:
    """EMI = [P * r * (1+r)^n] / [(1+r)^n – 1]"""
    r = annual_rate / (12 * 100)
    n = tenure_years * 12
    emi = principal * r * ((1 + r) ** n) / ((1 + r) ** n - 1)
    return round(emi, 2)


async def main():
    kernel = Kernel()

    # Register GPT-5 connector (works with your current SK version)
    kernel.add_service(
        OpenAIChatCompletion(
            ai_model_id="gpt-5",
            api_key=os.getenv("OPENAI_API_KEY"),
        )
    )

    # Register the function as a plugin
    kernel.add_plugin({"calculate_emi": calculate_emi}, plugin_name="FinanceTools")

    # ✅ Pass arguments directly as keyword arguments — SK will wrap them internally
    result = await kernel.invoke(
        plugin_name="FinanceTools",
        function_name="calculate_emi",
        principal=500000,
        annual_rate=10,
        tenure_years=3,
    )

    print(f"Calculated EMI: ₹{result}")


if __name__ == "__main__":
    asyncio.run(main())


Overwriting semantic_kernel_emi.py


In [91]:
!python semantic_kernel_emi.py

Calculated EMI: ₹16133.59


In [92]:
!pip show semantic-kernel

Name: semantic-kernel
Version: 1.37.0
Summary: Semantic Kernel Python SDK
Home-page: https://learn.microsoft.com/en-us/semantic-kernel/overview/
Author: 
Author-email: Microsoft <SK-Support@microsoft.com>
License: 
Location: /opt/anaconda3/lib/python3.11/site-packages
Requires: aiohttp, aiortc, azure-ai-agents, azure-ai-projects, azure-identity, cloudevents, defusedxml, jinja2, nest-asyncio, numpy, openai, openapi_core, opentelemetry-api, opentelemetry-sdk, prance, protobuf, pybars4, pydantic, pydantic-settings, scipy, typing-extensions, websockets
Required-by: 
